### ISSUE TITLE
Transparent objects brighten when using gammacorrectionshader

### ISSUE THREAD

**Issue by gloriousjob** [mrdoob/three.js#23019]
**Describe the bug**

When using sRGB textures (such as in 3DS models), the GammaCorrectionShader seems to brighten the texture more than when the EffectComposer is not used. The GammaCorrectionShader is needed when using the EffectComposer to produce similar colors to when the EffectComposer is not used. I have not been able to upgrade past r111 because of this issue.

As a side note, a similar problem was seen with generic shapes with color but it appears that setting the clearColor on the RenderPass can help with that issue but not with textured objects.

**To Reproduce**

1. Create a shape with a texture and an opacity of 0.4 and use an EffectComposer, with a GammaCorrectionShader pass.
2. Create a shape with an opacity of 0.4 just using a normal render (no EffectComposer).
3. Note that the colors are brighter on the first shape.

***Live example***

The following fiddle demonstrates the issue (top is no postprocessing, bottom is using postprocessing):
https://jsfiddle.net/s5tpxq1g/1/

You can see how the bottom is brighter for the texture and a lower opacity makes the issue more apparent.

**Expected behavior**

Colors match when using GammaCorrectionShader in an EffectComposer compared to not using EffectComposer.

**Screenshots**
Snapshot from fiddle (top half does not use EffectComposer, bottom half does). The left object shows that a normal object (with clearColor set correctly) does not have an issue but the right object, having textures, differs in brightness.:
![image](https://user-images.githubusercontent.com/9204009/145944427-805612f3-1d00-4738-9018-eb57684e5333.png)

**Platform:**

 - Device: Desktop
 - OS: Windows 10, 21H2
 - Browser: Chrome/Firefox/Edge
 - Three.js version: r134 but was noted in r112, when gammaInput and gammaOutput properties were removed and GammaCorrectionShader was recommended for sRGB scenarios.

**Comment by WestLangley** [comment]
The proper workflow is to render the scene in linear-sRGB color space and convert to sRGB color space as a final post-processing step.

However, currently, three.js converts to sRGB color space in-line in the shader, _prior_ to blending with the drawing buffer.

This is fine if the material is opaque. But if the material is transparent, it is not correct.

I would expect three.js to integrate a proper post-processing workflow into the engine at some point.

**Comment by bhouston** [comment]
We can also add a post processing step to convert a linear HDR RGB buffer to a LDR gamma corrected buffer.  You need at least an FP16 render target for this to work well.  It shouldn't really be hard -- I think all the code is in Three.js.  This is actually how Clara.io worked back in 2014: 

https://clara.io/view/9e2c46ce-6ada-450c-8acd-c44b84de2a38

I guess when you say we need to integrate a proper post-processing workflow, @WestLangley, you are meaning out of the box.  That makes sense now.

I actually switched from the proper HDR FP16/FP32 render buffer approach in 2016, and I contributed that back to Three.js here https://github.com/mrdoob/three.js/pull/8250, because many mobile devices simply didn't have enough memory to handle it properly, especially Apple iPads of the era.  The approach we switched to is what Three.js currently uses as the default workflow -- it was actually awesome and fast.  It does have a few limitations though as pointed out in this bug report.

Thus we could go back to this more traditional approach now that mobile hardware has caught up.

**Comment by bhouston** [comment]
And interesting @WestLangley and I had this exact same discussion back then: https://github.com/mrdoob/three.js/pull/8250#issuecomment-190445316

**Comment by Mugen87** [comment]
> However, currently, three.js converts to sRGB color space in-line in the shader, prior to blending with the drawing buffer.

With the current `dev` version and starting from `r136`, this does not happen with WebGL 2 anymore because the engine uses now `SRGB8_ALPHA8`.

**Comment by Mugen87** [comment]
So to clarify, the upper part of the fiddle is actually too dark whereas the lower part (using post-processing) is the correct image.

**Comment by bhouston** [comment]
@Mugen87 did you delete the comment where you mentioned the framebuffer/texture format "SRGB8_ALPHA8_EXT"?  Could we use this as the render target format?  From a quick reading, assuming the browser does the blending correctly between two separate sRGB8_ALPHA8_EXT buffers, it would solve the issue no?

**Comment by Mugen87** [comment]
Well, I was not sure about the details of my comment so I deleted it^^.

I think the goal should be to avoid all decodes/encodes and tone mapping in the shaders. Avoiding the decodes should be easier to implement via changes like #23039.

As mentioned earlier, avoiding the output encoding and tone mapping in GLSL means the renderer has to implement some sort of post-processing. I'm not yet sure how this would look like (API-wise) and I have not verified if `SRGB8_ALPHA8` could be used in this context but I _think_ it should work.

**Comment by looeee** [comment]
Note this has been [brought up](https://discourse.threejs.org/t/gamma-correction-on-transparent-objects/14527) [a few times over](https://discourse.threejs.org/t/gamma-correction-differences-when-object-is-transparent/14262) on the forum as well, so the current behavior does confuse people. 

The difference is also stronger the more transparent an object is - so there's not much difference with `opacity = 0.9` but with `opacity = 0.1` it's quite obvious.

**Comment by gloriousjob** [comment]
Not sure if I missed something but I tried updating the fiddle to use UnsignedByte, which should have enabled SRGB8_ALPHA  but it seems to have no difference:
https://jsfiddle.net/2qjth7un/

> So to clarify, the upper part of the fiddle is actually too dark whereas the lower part (using post-processing) is the correct image.

So when I've used things like a transparent vase, the post-processing makes it look like there's a light brightening the vase and makes seeing through the vase difficult so I think I'd actually prefer the upper picture, although I can see how the example I gave might look better.

**Comment by Mugen87** [comment]
> From a quick reading, assuming the browser does the blending correctly between two separate sRGB8_ALPHA8_EXT buffers, it would solve the issue no?

Yes. And according to https://github.com/KhronosGroup/WebGL/issues/2474#issuecomment-546595576, blending will indeed work correctly when using `sRGB8_ALPHA8`. 

However, I want to make clear that WebGL does not allow to define the color space of the default framebuffer. That means we have to perform the final linear-sRGB encode ourself in a custom pass.

I've hacked around and verified that the components of the framebuffer's color attachment are actually `gl.SRGB` when a render target has the RGBA8 format + `sRGBEncoding`. You can see this by checking the console output and the respective code in `WebGLTextures` in this demo:

https://raw.githack.com/Mugen87/three.js/dev19/examples/webgl_geometry_cube.html

As you can see, the scene is too dark since I have removed the `encodings_fragment` chunk from `MeshBasicMaterial`. Meaning no sRGB encode happens in the material shader anymore. The renderer has to integrate the FX pass and perform the sRGB encode in the shader.

**Comment by Kangz** [comment]
CCing @kainino0x. The group wants to have a way to configure the webgpu canvas with sRGB but it's not clear whether it will be by having `bgra8unorm` in the list of formats guaranteed to be supported (like is in the spec right now) or by allowing to configure the canvas with "format reinterpretation" so you can render to a `bgra8unorm-srgb` view into the `bgra8unorm` texture. See https://github.com/gpuweb/gpuweb/issues/1231

**Comment by kainino0x** [comment]
I am pretty sure that we will require applications to configure the canvas with, for example, `bgra8unorm` and then reinterpret to `bgra8unorm-srgb`:

> Don't allow `-srgb` formats as canvas texture formats. Instead, the texture _view_ needs to reinterpret the format as the `-srgb` equivalent. gpuweb/gpuweb#2336
> -- https://github.com/gpuweb/gpuweb/issues/1231#issuecomment-978868084

Unfortunately 2336 is still blocked in standardization but I'm certain we'll have **a way** to do this reinterpretation, so canvas can rely on it. (I don't want to add `bgra8unorm-srgb` as a valid canvas format because (1) it's redundant and (2) it is somewhat muddy whether the browser will sample it as `bgra8unorm` or `bgra8unorm-srgb` for compositing.)

In the meantime, you _should_ be able to create an intermediate `bgra8unorm-srgb` texture and `copyTextureToTexture` from it to the canvas (spec: https://github.com/gpuweb/gpuweb/pull/2329, Chromium: https://dawn-review.googlesource.com/c/dawn/+/74301).

**Comment by Mugen87** [comment]
I've updated `WebGLRenderer` to perform tone mapping and sRGB encoding as a separate post processing pass (https://github.com/Mugen87/three.js/commit/88549cf2dde1ffc0529f309b9b7b78597bff5fa8).

Unfortunately, a separate render pass noticeably downgrades the performance. Examples that used to run at 60 FPS on my iMac now run at approx. 40 FPS. Here are a live link for comparisons:

https://raw.githack.com/Mugen87/three.js/dev64/examples/index.html
https://threejs.org/examples/

If there is no way to improve the performance, it seems we have to keep using the inline approach for tone mapping and the sRGB encoding for the drawing framebuffer.

**Comment by bhouston** [comment]
The slowdown is to be expected.  I believe you now have a FP16 or FP32 linear render target right?  FP16 may lead to some speed improvements over FP32.   You could try that.  Just using that much memory could slow things down a lot, even without considering the separate pass.

And then a separate pass that reads this and writes to a sRGB output is another slow down.

I would suggest seeing if you wrote to the main linear render target as RGBA8 and had the separate pass the same as it is, what would the frame rate be?  I'm trying to separate out the cost of FP16, FP32 on the render time versus just the secondary pass architecture.

**Comment by bhouston** [comment]
I think it is reasonable to offer this separate pass mode as an option that people can enable.  Most people do not need it, thus most people can live with the high speed all in one mode, and only those who can, can enable this higher quality mode after the fact.

**Comment by Mugen87** [comment]
> I believe you now have a FP16 or FP32 linear render target right?

Yes, it's FP16. Testing with `webgl_animation_skinning_blending` and different types results in:

- RGBA8 ->47 FPS
- FP16 -> 45 FPS
- FP32 -> 39 FPS

Tested on an iMac with Chrome.

**Comment by bhouston** [comment]
So adding in the extra pass it reduces from 60FPS to 47FPS.  That is pretty significant, more than I thought it would.

I assume this isn't a new iMac, but rather an older one right?  How old?  What specific GPU/CPU?

**Comment by Mugen87** [comment]
> I assume this isn't a new iMac, but rather an older one right? How old? What specific GPU/CPU?

Yes, it is almost eight years old. Details:

- iMac (Retina 5K, 27'', ultimo 2014)
- 4 GHz Quad-Core Intel Core i7
- AMD Radeon R9 M295X 4 GB

**Comment by iuriiiurevich** [comment]
Here are some thoughts and some solution on this issue.

So the problem is that when rendering to the `WebGLRenderTarget` (in this case, using the `EffectComposer`), transparent objects mutate (opacity decreases, color becomes brighter).

As far as I understand, the problem is related to blending and is described as follows:
When the renderer renders the image directly onto the screen (in `sRGB`), everything is fine, because the mechanism is (roughly speaking):
1. Take all `linear` colors (decode those that are not linear)
2. Run these colors through shaders
3. Convert resulting colors to `sRGB` encoding
4. Blend them (considering the opacity)

Well done!

But when the renderer renders the image onto the `WebGLRenderTarget`, the mechanism is slightly different: step number 3 is missing, so that everything is blended in `linear` encoding and converted to `sRGB` later.

This workflow gives the correct result if all materials are opaque (no blending occurs between objects), but if there are some transparent objects, then obviously this will not give the same (correct) result.
So the problem is not actually related to the `GammaCorrectionShader`, because it works at the output of the `EffectComposer`, and the data is already incorrect at the input, immediately after the `RenderPass`.

The `SRGB8_ALPHA8` render target [doesn't solve this issue](https://github.com/mrdoob/three.js/issues/23019#issuecomment-1002236366) because as far as I understand `SRGB8_ALPHA8` is a storage format, so the blending of material shader results still occurs in `linear` color space before the data gets into this render target.

Is it right?

Anyway, there is one way to solve this problem: https://codesandbox.io/s/issue-23019-0wjd8n?file=/src/index.js
(had to slightly complicate [the original fiddle](https://github.com/mrdoob/three.js/issues/23019#issuecomment-1002236366) in order to demonstrate more complex cases)

Use `import { init } from "./wrong"` to see the issue and `import { init } from "./right"` to see the solution.

So the current solution is to blend objects in `sRGB` color space and write the result to the `EffectComposer`'s input buffer, and then, in its very first pass, convert the data to `linear` color space to work with it.
In order to blend objects in `sRGB` color space while writing to the render target, I had to hack into the Three code and allow it like so: https://github.com/mrdoob/three.js/pull/23870 (changed two lines of code).

So it takes an extra pass for the decoding, but this is hardly critical when using an `EffectComposer` with many passes.

@Mugen87 how bad is this solution?)
How could this problem be better solved?

Thanks!

**Comment by donmccurdy** [comment]
@iuriiiurevich I believe you've understood why EffectComposer works differently, but may have missed @WestLangley's comment above – the behavior with EffectComposer is correct, it's the default behavior in three.js that is (for most purposes) wrong.

I'm not sure there's anything we can change here... the 'inline' approach is dramatically cheaper, and it would be hard to justify removing the approach entirely. EffectComposer behaves correctly, and is available when the cost is acceptable. I would also point out that [pmndrs/postprocessing](https://github.com/pmndrs/postprocessing) has done a good thing in combining postprocessing passes — color space transforms, tone mapping, and other effects can be done in a single additional pass rather than two, which would be cheaper than implementing with three.js own EffectComposer today. I see @Mugen87 already combined two passes in his demo above.

tl;dr I would vote to close this. Per https://github.com/mrdoob/three.js/issues/23614#issuecomment-1143839598 their *might* be value in an 'unlit' color management workflow that doesn't go through Linear-sRGB, which would mean alpha blending happens in sRGB as a side effect. Beyond that (which I'm not even sure about) there doesn't seem to be anything we can change.

**Comment by iuriiiurevich** [comment]
@donmccurdy 
Thanks for the answer!
Summing up, what solutions can we offer to people who will face this issue?

As far as I understand, when a user is faced with the fact that the colors in Three.js (when using a Render Target) and, for example, in Photoshop [are blended differently](https://github.com/pmndrs/react-postprocessing/pull/108#issuecomment-1102576242), there are two main options:

1. Leave everything as it is, accept the difference in favor of performance.
2. Hack into Three.js' code using either @Mugen87's [approach](https://github.com/mrdoob/three.js/issues/23019#issuecomment-1019510488) or [this approach](https://github.com/mrdoob/three.js/issues/23019#issuecomment-1097520960) in favor of correct blending, but at the cost of performance.

Is it correct?
If so, shouldn't there be an easier way to achieve the correct color blending?

**Comment by donmccurdy** [comment]
To clarify one more thing, what I'm referring to as "correct" here is the behavior of blending Linear-sRGB components, instead of blending sRGB components. Photoshop has an option to enable/disable that, "Blend RGB Colors Using Gamma" — the preferred behavior may be different when working with 2D graphics, photography, 3D rendering, etc.

In a WebGL 3D renderer our goal is to blend colors in Linear-sRGB, and not sRGB<sup>1</sup>. Doing that requires post-processing, which can be expensive. Two cases:

1. With post-processing (using EffectComposer and GammaCorrectionShader), color is blended in Linear-sRGB, as it should be for a lit 3D rendering workflow.
2. Without post-processing, we simply can't do that, and color is blended in sRGB instead.

We cannot change (2), and we can't force all users to choose post-processing, but either way no hacks are required. three.js provides an API to apply the view transform (tone mapping and color space transform) in post-processing, users can enable that if the performance cost is acceptable. Some applications will need post-processing for other reasons.

***

<sup>1</sup> To be fair, some developers do use three.js for working with 2D graphics and unlit workflows, and they might prefer blending in sRGB. That's fine, and I wish we could offer a simple toggle for this, but technically our hands are tied.

**Comment by Mugen87** [comment]
Minot update: This is how the fiddle should look like with the upcoming `r153`: https://jsfiddle.net/t902fnr8/

sRGB and color management are defaults now as well as the new `OutputPass` (which combines tone mapping and color space conversion).

**Comment by Mugen87** [comment]
As pointed out in https://github.com/mrdoob/three.js/issues/23019#issuecomment-1166308009 and https://github.com/mrdoob/three.js/issues/23019#issuecomment-1175096174, this issue can be closed. 

We might want to note the usage of the new `OutputPass` in the post processing or color management guide though so it's more visible for the user.
